contracts/utils/EnumerableSet.sol

```diff
@@ -53,10 +53,19 @@ library EnumerableSet {
      */
     function _add(Set storage set, bytes32 value) private returns (bool) {
         if (!_contains(set, value)) {
-            set._values.push(value);
-            // The value is stored at length-1, but we add 1 to all indexes
-            // and use 0 as a sentinel value
-            set._indexes[value] = set._values.length;
+            uint256 last;
+            assembly {
```

Comment:
I feel too uneasy about this assembly block. I prefer to have an extra `sload`.

If we want to insist on the optimization, we should use as little assembly as possible:

```solidity
    function push(bytes32[] storage array, bytes32 value) internal returns (uint256) {
        uint arraySlot;
        uint[1] storage arrayLength;
        assembly {
            arraySlot := array.slot
            arrayLength.slot := array.slot
        }
        uint arrayLastIndex = arrayLength[0];
        arrayLength[0] += 1;
        uint arrayLastSlot = arrayLastIndex + uint256(keccak256(abi.encodePacked(arraySlot)));
        bytes32[1] storage arrayLast;
        assembly {
            arrayLast.slot := arrayLastSlot
        }
        arrayLast[0] = value;
        return arrayLastIndex;
    }
```

However, I still worry about the clarity of all this.

Reply:
yes, clarity is not great

Reply:
Then I think we should keep the previous code even if there's a redundat `SLOAD` of the length.